source/extensions/network/dns_resolver/cares/dns_impl.cc

```diff
@@ -507,6 +508,11 @@ class CaresDnsResolverFactory : public DnsResolverFactory {
                                              typed_dns_resolver_config) const override {
     envoy::extensions::network::dns_resolver::cares::v3::CaresDnsResolverConfig cares;
     std::vector<Network::Address::InstanceConstSharedPtr> resolvers;
+    // Initialize c-ares library in case first time.
+    if (!ares_library_initialized_) {
+      ares_library_init(ARES_LIB_INIT_ALL);
+      ares_library_initialized_ = true;
```

Comment:
what threads can this be called on? Consider Envoy-Mobile case where there can be multiple main threads.

Reply:
Good point. Let me move the ares_library_init() to main thread. Please check the latest change, which call the ares_library_init() within factory. init() after factory retrieval. And IIUC,  the factory retrieval is done in maint hread.
https://github.com/envoyproxy/envoy/blob/326b205af6b16dbf75f6bfbf246082f5966bc0b0/source/common/network/dns_resolver/dns_factory_util.cc#L82

BTW, createDnsResolver() can be called in both main threads, like here:
https://github.com/envoyproxy/envoy/blob/326b205af6b16dbf75f6bfbf246082f5966bc0b0/source/server/server.cc#L681

Or worker threads, like here
https://github.com/envoyproxy/envoy/blob/326b205af6b16dbf75f6bfbf246082f5966bc0b0/source/extensions/filters/udp/dns_filter/dns_filter_resolver.h#L28

I tried to move the latter to main thread, but later on found doing it in worker threads is actually by design. Please check : https://github.com/envoyproxy/envoy/pull/21887#issuecomment-1166283363.